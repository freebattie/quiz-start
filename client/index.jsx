import { createRoot, useEffect } from "react-dom/client";
import {
  BrowserRouter,
  Routes,
  Route,
  Link,
  createRoutesFromChildren,
} from "react-router-dom";
import { useState } from "react";

// get the root element in index.html
const container = document.getElementById("app");
//creat a react root from that id
const root = createRoot(container);

function StartPage({ totalQ, correct }) {
  return (
    <div>
      <h1>Welcome too the start page</h1>
      <h2>
        You have answered {totalQ} {totalQ > 1 ? "questions" : "question"}
      </h2>
      <h2>and have {correct} correct</h2>
      <ul>
        <li>
          <Link to={"/quiz"}>Take Quiz</Link>
        </li>
        <li>
          <Link to={"/quiz/new"}> Add a Quiz</Link>
        </li>
      </ul>
    </div>
  );
}

function ShowError() {
  return <h1>ERROR PAGE</h1>;
}

function ShowQuiz() {
  return <h1>THis is the QUIZ</h1>;
}

function AddQuiz() {
  return null;
}

function QuzApp() {
  return (
    <Routes>
      <Route path={"/"} element={<ShowQuiz />} />
      <Route path={"/new"} element={<AddQuiz />} />
      <Route path={"*"} element={<ShowError />} />
    </Routes>
  );
}

function Application() {
  const [correct, setCorrect] = useState(0);
  const [totalQ, setTotalQ] = useState(0);
  return (
    <BrowserRouter>
      <Routes>
        <Route
          path={"/"}
          element={<StartPage correct={correct} totalQ={totalQ} />}
        />
        <Route
          path={"/quiz/*"}
          element={<QuzApp onAddScore={setCorrect} onAddTotal={setTotalQ} />}
        />
      </Routes>
    </BrowserRouter>
  );
}

// render a h1 tag with "hello world"
root.render(<Application />);
