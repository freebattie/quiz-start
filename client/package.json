{
  "name": "client",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "dev": "parcel watch index.html"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "^7.18.6",
    "babel-jest": "^29.0.3",
    "jest": "^29.0.3",
    "parcel": "^2.7.0",
    "rimraf": "^3.0.2"
  },
  "dependencies": {
    "@parcel/transformer-react-refresh-wrap": "^2.7.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.0"
  },
  "description": ""
}
