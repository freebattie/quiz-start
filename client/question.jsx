const quizzes = [
  {
    question: "What is the capital of Norway?",
    answers: ["Oslo", "Stavanger", "Bergen", "Kristiansand"],
    correct: 0,
  },
  {
    question: "What is the capital of Sweden?",
    answers: ["Oslo", "Stockholm", "Bergen", "Kristiansand"],
    correct: 1,
  },
  {
    question: "What is the largest city in Norway",
    answers: ["Ålesund", "Stavanger", "Oslo", "Kristiansand"],
    correct: 2,
  },
  {
    question: "What is not a city in Norway",
    answers: ["Oslo", "Stavanger", "Gøteborg", "Kristiansand"],
    correct: 2,
  },
  {
    question: "What is country dose not share a land border with norway?",
    answers: ["Russia", "Sweden", "Finland", "Denmark"],
    correct: 3,
  },
];
export function addQuiz(quiz) {
  quizzes.push(quiz);

  quizzes.forEach((e) => console.log(e));
}
export function isCorrectAnswer(index, answer) {
  return quizzes[index].answers.indexOf(answer) === quizzes[index].correct;
}

function QuestionCard({ quiz }) {
  return (
    <div>
      <h1>{quiz.question}</h1>
      {quiz.answers.map((a, key) => (
        <div key={key}>{a}</div>
      ))}
    </div>
  );
}

export function GetAllQuestions(event) {
  return quizzes.map((quiz, key) => {
    return <QuestionCard key={key} quiz={quiz} />;
  });
}
export function getRandomQuestion() {
  const newQ = Math.trunc(Math.random() * quizzes.length);

  return {
    index: newQ,
    question: quizzes[newQ].question,
    answers: quizzes[newQ].answers,
  };
}
