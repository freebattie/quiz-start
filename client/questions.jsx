const quizzes = [
  {
    question: "What is the capital of Norway?",
    answers: ["Oslo", "Stavanger", "Bergen", "Kristiansand"],
    correct: 0,
  },
  {
    question: "What is the capital of Sweden?",
    answers: ["Oslo", "Stockholm", "Bergen", "Kristiansand"],
    correct: 1,
  },
  {
    question: "What is the largest city in Norway",
    answers: ["Ålesund", "Stavanger", "Oslo", "Kristiansand"],
    correct: 2,
  },
  {
    question: "What is not a city in Norway",
    answers: ["Oslo", "Stavanger", "Gøteborg", "Kristiansand"],
    correct: 2,
  },
  {
    question: "What is country dose not share a land border with norway?",
    answers: ["Russia", "Sweden", "Finland", "Denmark"],
    correct: 3,
  },
];
export function addQuiz(quiz) {
  quizzes.push(quiz);

  quizzes.forEach((e) => console.log(e));
}
export function isCorrectAnswer(index, answer) {
  return quizzes[index].answers.indexOf(answer) === quizzes[index].correct;
}

export function questions() {
  const newQ = Math.trunc(Math.random() * quizzes.length);

  return {
    index: newQ,
    question: quizzes[newQ].question,
    answers: quizzes[newQ].answers,
  };
}
