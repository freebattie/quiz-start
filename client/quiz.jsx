import {
  addQuiz,
  GetAllQuestions,
  isCorrectAnswer,
  getRandomQuestion,
} from "./question.jsx";
import { Route, Routes, useLocation, useNavigate } from "react-router-dom";
import { createContext, useContext, useEffect, useState } from "react";

function ShowError() {
  return <GetAllQuestions />;
}
export const QuestionContext = createContext({ getRandomQuestion });
export function ShowQuiz({ onAddScore, onAddTotal }) {
  const wrong = "/quiz/answer/wrong";
  const correct = "/quiz/answer/correct";
  function handelClick(answer) {
    onAddTotal((t) => t + 1);
    if (isCorrectAnswer(index, answer)) {
      onAddScore((s) => s + 1);
      navigate(correct, { state: { type: "CORRECT" } });
    } else {
      navigate(wrong, { state: { type: "WRONG" } });
    }
  }

  const { getRandomQuestion } = useContext(QuestionContext);
  const [rndQ] = useState(getRandomQuestion());
  const navigate = useNavigate();
  const [index, setIndex] = useState(rndQ.index);
  const [question, setQuestion] = useState(rndQ.question);
  const [answers, setAnswers] = useState(rndQ.answers);

  return (
    <div>
      <div>{question}</div>
      {answers.map((a) => {
        return (
          <button key={a} onClick={() => handelClick(a)}>
            {a}
          </button>
        );
      })}
    </div>
  );
}

export function AddQuiz() {
  const navigate = useNavigate();
  const [question, setQuestion] = useState("Bjarte");
  const [answer1, setAnswer1] = useState("answer1");
  const [answer2, setAnswer2] = useState("answer2");
  const [answer3, setAnswer3] = useState("answer3");
  const [answer4, setAnswer4] = useState("answer4");
  const [correct, setCorrect] = useState(1);
  const [newQuiz, setNewQuiz] = useState({
    question: "",
    answers: ["one", "two", "three", "four"],
    correct: 0,
  });
  useEffect(() => {
    setNewQuiz({
      question,
      answers: [answer1, answer2, answer3, answer4],
      correct,
    });
  }, [question, answer1, answer2, answer3, answer4, correct]);

  function handelOnSubmit(event) {
    event.preventDefault();
    addQuiz(newQuiz);

    navigate("..");
  }

  return (
    <form onSubmit={handelOnSubmit}>
      <h1>create new movie question</h1>
      <div>
        Question:
        <input
          value={question}
          onChange={(e) => {
            setQuestion(e.target.value);
          }}
        />
      </div>
      <div>
        Answer1:
        <input
          value={answer1}
          onChange={(e) => {
            setAnswer1(e.target.value);
          }}
        />
      </div>
      <div>
        Answer2:
        <input
          value={answer2}
          onChange={(e) => {
            setAnswer2(e.target.value);
          }}
        />
      </div>
      <div>
        Answer3:
        <input
          value={answer3}
          onChange={(e) => {
            setAnswer3(e.target.value);
          }}
        />
      </div>
      <div>
        Answer4:
        <input
          value={answer4}
          onChange={(e) => {
            setAnswer4(e.target.value);
          }}
        />
      </div>
      <div>
        Correct:
        <input
          type="number"
          min="0"
          max="4"
          value={correct}
          onChange={(e) => {
            if (e.target.value >= 4) {
              setCorrect(3);
            } else {
              if (e.target.value !== "") {
                setCorrect(parseInt(e.target.value));
              } else setCorrect(e.target.value);
            }
          }}
        />
      </div>
      <button>Save</button>
      <pre>{JSON.stringify(newQuiz)}</pre>
    </form>
  );
}

function ReturnAnswer() {
  const navigate = useNavigate();
  const { state } = useLocation();
  if (state != null) {
    const { type } = state;
    return (
      <div>
        <h1>{type}</h1>
        <button
          onClick={() => {
            navigate("/");
          }}
        >
          BACK
        </button>
      </div>
    );
  } else return <ShowError />;
}

export function QuzApp({ onAddScore, onAddTotal }) {
  return (
    <Routes>
      <Route
        path={"/"}
        element={<ShowQuiz onAddScore={onAddScore} onAddTotal={onAddTotal} />}
      />
      <Route path={"new"} element={<AddQuiz />} />
      <Route path={"answer/:path"} element={<ReturnAnswer />} />
      <Route path={"*"} element={<ShowError />} />
    </Routes>
  );
}
