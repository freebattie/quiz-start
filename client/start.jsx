import { Link } from "react-router-dom";

export function StartPage({ totalQ, correct }) {
  return (
    <div>
      <h1>Welcome too the start page</h1>
      <h2>
        You have answered {totalQ} {totalQ > 1 ? "questions" : "question"}
      </h2>
      <h2>and have {correct} correct</h2>
      <ul>
        <li>
          <Link to={"/quiz"}>Take Quiz</Link>
        </li>
        <li>
          <Link to={"/quiz/new"}> Add a Quiz</Link>
        </li>
      </ul>
    </div>
  );
}
