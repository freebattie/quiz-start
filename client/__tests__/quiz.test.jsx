import * as React from "react";

import { createRoot } from "react-dom/client";

import { act } from "react-dom/test-utils";
import { StartPage } from "../start.jsx";
import { GetAllQuestions } from "../questions.jsx";
global.IS_REACT_ACT_ENVIRONMENT = true;
describe("quiz page", function () {
  it("should show QuizPage", function () {
    const element = document.createElement("div");

    //creat a React root from that id

    act(() => {
      const root = createRoot(element);
      root.render(<GetAllQuestions />);
    });

    expect(element).toMatchSnapshot();
  });
});
