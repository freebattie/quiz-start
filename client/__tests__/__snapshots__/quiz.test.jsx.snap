// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quiz page should show QuizPage 1`] = `
<div>
  <div>
    <h1>
      What is the capital of Norway?
    </h1>
    <div>
      Oslo
    </div>
    <div>
      Stavanger
    </div>
    <div>
      Bergen
    </div>
    <div>
      Kristiansand
    </div>
  </div>
  <div>
    <h1>
      What is the capital of Sweden?
    </h1>
    <div>
      Oslo
    </div>
    <div>
      Stockholm
    </div>
    <div>
      Bergen
    </div>
    <div>
      Kristiansand
    </div>
  </div>
  <div>
    <h1>
      What is the largest city in Norway
    </h1>
    <div>
      Ålesund
    </div>
    <div>
      Stavanger
    </div>
    <div>
      Oslo
    </div>
    <div>
      Kristiansand
    </div>
  </div>
  <div>
    <h1>
      What is not a city in Norway
    </h1>
    <div>
      Oslo
    </div>
    <div>
      Stavanger
    </div>
    <div>
      Gøteborg
    </div>
    <div>
      Kristiansand
    </div>
  </div>
  <div>
    <h1>
      What is country dose not share a land border with norway?
    </h1>
    <div>
      Russia
    </div>
    <div>
      Sweden
    </div>
    <div>
      Finland
    </div>
    <div>
      Denmark
    </div>
  </div>
</div>
`;

exports[`quiz page should show answer status 1`] = `
"<div>
  <h1>Welcome too the start page TEST</h1>
  <h2 data-testid="total">You have answered 8 questions</h2>
  <h2 data-testid="correct">and have 3 correct</h2>
  <ul>
    <li><a href="/quiz">Take Quiz</a></li>
    <li><a href="/quiz/new"> Add a Quiz</a></li>
  </ul>
</div>"
`;

exports[`quiz page should show question 1`] = `
"<div>
  <h1>What is the capital of Norway?</h1>
  <div>Oslo</div>
  <div>Stavanger</div>
  <div>Bergen</div>
  <div>Kristiansand</div>
</div>
<div>
  <h1>What is the capital of Sweden?</h1>
  <div>Oslo</div>
  <div>Stockholm</div>
  <div>Bergen</div>
  <div>Kristiansand</div>
</div>
<div>
  <h1>What is the largest city in Norway</h1>
  <div>Ålesund</div>
  <div>Stavanger</div>
  <div>Oslo</div>
  <div>Kristiansand</div>
</div>
<div>
  <h1>What is not a city in Norway</h1>
  <div>Oslo</div>
  <div>Stavanger</div>
  <div>Gøteborg</div>
  <div>Kristiansand</div>
</div>
<div>
  <h1>What is country dose not share a land border with norway?</h1>
  <div>Russia</div>
  <div>Sweden</div>
  <div>Finland</div>
  <div>Denmark</div>
</div>"
`;
